packages/salesforcedx-vscode-core/src/commands/forceSourceDiff.ts

```diff
@@ -0,0 +1,159 @@
+
+export class ForceSourceDiffExecutor extends SfdxCommandletExecutor<{
+  filePath: string;
+}> {
+  public build(data: { filePath: string }): Command {
+    const commandBuilder = new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_source_diff_text'))
+      .withArg('force:source:diff')
+      .withLogName('force_source_diff')
+      .withFlag('--sourcepath', data.filePath)
+      .withJson();
+    return commandBuilder.build();
+  }
+
+  public async execute(
+    response: ContinueResponse<{ filePath: string }>
+  ): Promise<void> {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+
+    channelService.streamCommandStartStop(execution);
+    channelService.showChannelOutput();
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(realData => {
+      stdOut += realData.toString();
+    });
+
+    execution.processExitSubject.subscribe(async () => {
+      this.logMetric(execution.command.logName, startTime);
+
+      try {
+        const diffParser = new DiffResultParser(stdOut);
+        const diffParserSuccess = diffParser.getSuccessResponse();
+        const diffParserError = diffParser.getErrorResponse();
+
+        if (diffParser.isSuccessful() && diffParserSuccess) {
+          const diffResult = diffParserSuccess.result;
+          const remote = vscode.Uri.parse(diffResult.remote);
+          const local = vscode.Uri.parse(diffResult.local);
+          const filename = diffResult.fileName;
+
+          let defaultUsernameorAlias: string | undefined;
+          if (hasRootWorkspace()) {
+            defaultUsernameorAlias = await OrgAuthInfo.getDefaultUsernameOrAlias(
+              false
+            );
+          }
+          vscode.commands.executeCommand(
+            'vscode.diff',
+            remote,
+            local,
+            nls.localize(
+              'force_source_diff_title',
+              defaultUsernameorAlias,
+              filename,
+              filename
+            )
+          );
+        } else if (diffParserError) {
+          channelService.appendLine(diffParserError.message);
+        }
+      } catch (e) {
+        telemetryService.sendException(e.name, e.message);
+        const err = new Error('Error parsing diff result');
+        err.name = 'DiffParserFail';
+        throw err;
+      }
+    });
+
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+}
+
+const workspaceChecker = new SfdxWorkspaceChecker();
+
+export class SourcePathGatherer
```

Comment:
[FilePathGatherer](https://github.com/forcedotcom/salesforcedx-vscode/blob/b5af214c2421f6c916f5e11d2140ffdbec62f067/packages/salesforcedx-vscode-core/src/commands/commands.ts#L115) already exists in `commands.ts`